Step1 : 레거시 코드 리팩터링

Mo-bile opened the pull request:
안녕하세요? 모재영입니다 👍 

점점 실무 코드와 가까워 지는것을 보면서 무언가 어서 적용해보고 싶다는 생각이 든 과제였습니다
좋은 과제 감사드립니다.
처음에는 대체 이걸 어떻게 해야하나 싶어서 조금씩 만지다가 Answer 을 먼저 테스트하고, 도메인으로 분리하고 리팩토링 하고 그 결과를 보면서 오랜만에 코딩에서 재미를 느끼게 되었습니다. 감사합니다 😊

하나 궁금한점이 있습니다 `규칙7 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않기` 에 의거해 
`CreateAnswerCommand`, `CreateDeleteHistoryCommand`, `CreateQuestionCommand` 3개의 클래스를 만들어 분리했습니다.
사실 만들고나면서 생각든게 단순 분리이고 따로 테스트가 추가된다거나, 분리 후 상위 클래스와의 결합으로 특별히 필요한 메서드는 보이지 않았습니다.
혹시 이렇게 시도하는 것 맞을까요?

감사합니다.

Commit:
refactor : List<Answer> 부분 일급 컬렉션으로 Answers 전환 및 비즈니스 로직 이관

Commit:
doc : 리팩터링 절차 및 계획안 마련

Commit:
refactor
1. DeleteHistories 로 일급컬렉션 뽑아내고, 관련 로직 도메인으로 이관
2. answer 제거 및 history 저장관련 내용 메서드 분리

Commit:
refactor
1. 메서드명 수정하여 그 의미 명확화
2. DeleteHistories 생성자 추가하여 생성 역할 다양화
3. deleteQuestion 메서드 리팩터링
4. Question의 일부 비즈니스 로직 도메인 내로 이관

Commit:
refactor
1. 불필요한 setter 제거

Commit:
refactor : 기존적인 생성, 수정일시 관련 공통된 인스턴스 필드 분리

Commit:
docs : 리팩토링 상황 메모

Commit:
refactor : 규칙 7 인스턴스 변수 줄이기

Commit:
refactor : 컨벤션 및 import 정리

Commit:
refactor : 테스트 픽스쳐 청리

javajigi commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -57,23 +45,27 @@ public boolean isDeleted() {
     public boolean isOwner(NsUser writer) {
-        return this.writer.equals(writer);
+        return this.createAnswerCommand.isOwner(writer);
     }
```
현재 구현한 방식과 아래와 같이 글쓴이 인지 여부를 이 메서드가 판단하도록 구현하는 것 중 어느 방식으로 구현하는 것이 좋을지 고민해 본다.
위 두 가지 방식 중 선택한 방법으로 구현하는 것이 좋은 이유에 대해 피드백으로 남겨본다.

```java
    public void delete(NsUser loginUser){
        // 글쓴이 인지를 판단하는 로직 검증 후 삭제 상태로 변경
    }
```

Mo-bile replied:
1. 도메인 불변식 유지 : 모델이 깨지면 안되는 규칙 -> 어떤 경로이든 참이어야 함
    - 규칙검증을 도메인 밖인 서비스에 두면, 제3의 다른서비스는 이 규칙 검증을 건너뛸 수있게 됨
    - 규칙이 도메인 밖으로 새면 안됨 -> 일관된 규칙 보호가 가능해지게 됨

2. 서비스 레이어 메서드는 외부(DB, I/O, 셔플)와 `@Transactional` 가 관심사임
    - 비즈니스 규칙 자체는 도메인 내부에서 해결해야함

javajigi commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,73 @@
+
+public class Answers {
+    
+    private final List<Answer> answers;
```
일급 콜렉션 적용 👍

javajigi commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,73 @@
+            .allMatch(Answer::isDeleted);
+    }
+    
+    public boolean isAllDelete() {
```
```suggestion
    public boolean deleteAll(NsUser loginUser) {
       ...
    }
   
```
앞의 Answer의 delete 피드백과 같음

Mo-bile replied:
이유
도메인 불변식 유지 & 서비스 레이어는 다른 외부 서비스, DB 등과의 조율이 주요 목표

javajigi commented on src/main/java/nextstep/qna/domain/CreateAnswerCommand.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Objects;
+import nextstep.users.domain.NsUser;
+
+public class CreateAnswerCommand {
```
인스턴스 변수를 줄이기 위한 약간은 과하면서도 무리한 객체 분리로 느껴짐
각 인스턴스 변수가 서로 관련성이 있는 경우 분리하는 것이 의미있지 지금과 같이 분리하는 것은 의미가 없을 것으로 판단됨

Mo-bile replied:
객체 분리 필요에 대해서 고민한 결과
분리 된 객체만의 특별한 규칙, 행동이 없다.
그 결과 이 객체는 단순히 보관하는 DTO 의 역할에 불과하다.

javajigi commented on src/main/java/nextstep/qna/domain/CreateDeleteHistoryCommand.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Objects;
+import nextstep.users.domain.NsUser;
+
+public class CreateDeleteHistoryCommand {
```
CreateAnswerCommand와 같은 피드백

Mo-bile replied:
마찬가지로 분리된 객체만의 특별한 규칙과 행동이 없다. DTO 에 불과함

javajigi commented on src/main/java/nextstep/qna/domain/CreateQuestionCommand.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.Objects;
+import nextstep.users.domain.NsUser;
+
+public class CreateQuestionCommand {
```
CreateAnswerCommand와 같은 피드백

Mo-bile replied:
단순히 값을 보관해주는 DTO 그 이상의 의미를 찾기가 어려움

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -32,61 +22,80 @@ public Question(NsUser writer, String title, String contents) {
-    public Question setDeleted(boolean deleted) {
-        this.deleted = deleted;
+    
+    public boolean delete() {
```
현재 구현한 방식과 아래와 같이 글쓴이 인지 여부를 이 메서드가 판단하도록 구현하는 것 중 어느 방식으로 구현하는 것이 좋을지 고민해 본다.
위 두 가지 방식 중 선택한 방법으로 구현하는 것이 좋은 이유에 대해 피드백으로 남겨본다.

```java
    public void delete(NsUser loginUser){
        // 글쓴이 인지를 판단하는 로직 검증 후 삭제 상태로 변경
    }
```

Mo-bile replied:
삭제시 도메인 내부에 글쓴이 인지 판단하는 로직이 들어가야한다.
그 이유는 제3의 서비스에서 삭제시 글쓴이 인지 판단하는 여부를 누락할 수있다.

javajigi commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -1,49 +1 @@
        }
        return List.of();
    }
}
```
코드 깨짐 확인 바람

javajigi reviewed (changes requested):
1단계 미션 진행 👍 
"처음에는 대체 이걸 어떻게 해야하나 싶어서 조금씩 만지다가 Answer 을 먼저 테스트하고, 도메인으로 분리하고 리팩토링 하고 그 결과를 보면서 오랜만에 코딩에서 재미를 느끼게 되었습니다."
이런 느낌을 받았다니 다행이네요.
우리가 흔히 접하는 웹 애플리케이션에서도 OOP 기반으로 개발하고 TDD 적용이 가능한 만큼 도전해 볼 것을 추천해요.

"규칙7 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않기" 원칙을 적용한 것은 넘 좋은데요.
이 원칙을 현장에 적용하는데는 어려움이 있더라고요.
너무 무리한 원칙 적용보다는 꼭 필요한 곳이 있을지 찾아 적용해 볼 것을 추천합니다.

Commit:
refactor : 깨진 코드 재 커밋

Commit:
refactor : 깨진 코드 재 커밋2

Commit:
refactor : 깨진 코드 해결용 커밋

Mo-bile commented:
안녕하세요?
출근 길에 아침에 일어나서 `QnAService` 가 깨져있어 놀랬습니다.
인텔리j 는 문제가 없었는데 github 내 에서는 `QnAService` 만 깨져있네요 원인을 알아보니 line seperator 설정이 문제가 되어 코드가 깨진것 같습니다.  푸시 후 file changes 를 확인을 다 했어야 했는데 미쳐 하지못한 점 양해 해부탁드립니다.

다른 부분을 고치고 리뷰 요청드릴까 했지만 `QnAService` 가 핵심 부분 중 하나이다 보니 이 코드까지 리뷰해주신 뒤 하면 좋을 것 같아서 바로 리뷰 요청드렸습니다.
감사힙니다 👍

javajigi commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -1,49 +1,52 @@
+        
+        Answers answers = question.sameUserQuestionAndAnswer();
+        
+        deleteHistoryService.saveAll(new DeleteHistories(deleteQuestion(question), deleteAnswers(answers)).toList());
```
```suggestion
        DeleteHistories deleteHistories = question.delete(loginUser);
        deleteHistoryService.saveAll(deleteHistories);
```
Answers를 question에서 get한 후 구현할 필요가 있을까?
위와 같이 question에게 delete() 하라는 메시지를 보내는 것만으로 구현할 수 있지 않을까?

Mo-bile replied:
서비스 레이어의 `deleteHistory()`는 질문을 삭제하기를 원할 뿐이지
`Answers`를 꺼내서 한다면 제3의 서비스 레이어에서 `Answers` 제거를 누락할 수있다.

javajigi reviewed (changes requested):
QnaService도 피드백 남겼으니 함께 반영해 보세요.

Commit:
refactor
1. 도메인 규칙은 해당하는 도메인 내부에 캡슐화 시켜서 도메인 불변식을 유지한다. 그 결과 해당 도메인을 모르는 제3의 서비스에서 삭제를 할 때 삭제 규칙을 누락하는 경우를 방지한다.
2. `규칙7 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않기` 를 무리하게 이용된 부분 제거

Commit:
doc : 리팩터링 진행 상황 메모

Mo-bile commented:
안녕하세요?
이전에 꼼꼼한 리뷰 감사드립니다. 👍 
해주신 리뷰에서 그 이유에 대해서 고민해보게 하고, 남겨보게 해주셔서 감사합니다.

그 덕에 평소 보다 더 긴 시간 깊게 고민하고 글로 남기면서 말씀 주신 방향에 대해서 더욱 납득하게 되었습니다.

Commit:
refactor : deleteAll, delete 메서드의 빈 리스트, 클래스 반환하여 하는 방법이 아닌, 예외 전파방식으로 무의미한 DeleteHistories 생성 방지

javajigi commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,80 @@
+    
+    public List<DeleteHistory> deleteAll(NsUser loginUser) throws CannotDeleteException {
+        isHaveAuthority(loginUser);
+        arsAnswersDeleteDone();
```
```suggestion
        for(Answer answer: this.answers) {
            answer.delete(loginUser);
        }
```

answer에 대해서도 같은 방식으로 구현해야하지 않을까?
지금과 같이 구현하는 것은 answer를 수동적인 객체로 바라보는 느낌이 듦

Mo-bile replied:
answers 를 일급컬렉션으로 만드는것에 집중한 나머지 answer 을 수동적으로 바라보아
삭제 권한 검증을 answers 에서 하게되고, 그와 별도로 삭제 자체도 answers 에서 시도하게 되었습니다.

answer 에서 필요한이유는
1. answers 에 검증과 삭제가 있으면 answer 은 getter만 남는 자기 행동이 없는 수동적인 객체가 되는 점
2. 또한 answers 가 answer 의 삭제 로직을 대신 수행하게 되는 점
때문입니다.

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -42,51 +36,86 @@ public Long getId() {
 
     public String getTitle() {
-        return title;
+        return this.getTitle();
```
```suggestion
        return this.title;
```

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -42,51 +36,86 @@ public Long getId() {
+        return this.getTitle();
     }
 
     public Question setTitle(String title) {
```
Question, Answer의 모든 setter 메서드 제거해도 되지 않을까?

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -42,51 +36,86 @@ public Long getId() {
+    public DeleteHistories delete(NsUser loginUser) throws CannotDeleteException {
+        this.isHaveAuthority(loginUser);
+        this.deleteQuestion();
+        this.isQuestionDeleteDone();
```
이 부분에 대한 확인이 굳이 필요할까?
불필요한 확인으로 보여짐

Mo-bile replied:
말씀주신 바 대로 고민을 해보니 필요가 없다고 생각합니다
그 이유는 `this.isQuestionDeleteDone(); ` 같은 도메인 내부에서 값의 변화를 검증하는 부분인데, 

1. 도메인 내부 값 변화는 확실한 점
2. 그래서 도메인 내부에서 값 변화를 체크하는 방어코드는 무의미한 점 (Anti-pattern)
3. 만약 검증하려면 db(영속성) 레이어 혹은 트랜잭셔널 범위(서비스 레이어) 등에서 검증해야하지만 금번 과제는 리팩터링 이므로 추가 할 필요가 없는점 
이므로 제거 하겠습니다

javajigi commented on src/test/java/nextstep/qna/domain/AnswerTest.java:
```diff
@@ -3,6 +3,7 @@
 import nextstep.users.domain.NsUserTest;
 
 public class AnswerTest {
```
Answer에 대한 삭제에 대한 테스트도 있어야 하지 않을까?

javajigi reviewed (changes requested):
바로 머지하려다 약간의 아쉬움이 남는 부분이 있어 피드백 남겨봤어요. 👍 
고민이 많았다고 했는데요.
남긴 피드백을 통해 좀 더 생각해 보는 시간을 가지면 좋겠습니다.

Commit:
refactor : 무의미한 방어코드를 제거(도메인 내부는 무조건 true / 실익은 영속성 레이어, 트랜잭셔널 범위에서 검증에서)

Commit:
refactor :
1. Answer 을 수동적으로 바라보아서 Answer에 대한 검증을 Answers 에서 하게한 것을 Answer로 이관
2. 불필요한 메서드 및 테스트 코드 제거

Commit:
doc : 이유 작성

Mo-bile commented:
안녕하세요? 꼼꼼히 주신 피드백 덕분에 무릎을 여러번 쳤습니다 감사합니다 👍 
하나 질문드려도 될까요?
`Answers` 의 deleteAll() 에 answer의 `isHaveAuthority()` 을 반복문으로 호출해야할까 고민이됩니다.

그 이유는 다음과 같습니다
`Answers` 의 deleteAll() 로 `answer.delete()` 를 호출하는데요
`answer.delete()` 내부에 '질문자와 답변자 동일 검증' 과 '삭제'를 호출 했습니다.

```java
// Answers 내부
public List<DeleteHistory> deleteAll(NsUser loginUser) throws CannotDeleteException {
    deleteAllAnswers(loginUser);
    return addInDeleteHistory();
}

private void deleteAllAnswers(NsUser loginUser) throws CannotDeleteException {
    for(Answer answer: answers) {
        answer.delete(loginUser);
    }
}

// Answer 내부
public void delete(NsUser loginUser) throws CannotDeleteException {
    isHaveAuthority(loginUser);
    deleteAnswer();
}
```

그런데 이 방식을 생각해보니
첫 번째 answer → owner = userA → 삭제 성공 (deleted = true)
두 번째 answer → owner = userA → 삭제 성공
세 번째 answer → owner = userB → CannotDeleteException 발생
그 결과 예외 던져졌지만 앞의 2개는 이미 deleted = true로 바뀐 상태로 볼 수있습니다.

즉 반쯤 삭제하다가 중간에 예외가 나는데요 이렇게 접근해도 괜찮을까 생각들었습니다.

그랬다가 피드백 요청 직전 커밋 상태 유지로 결론냈습니다.
1. answer 객체의 도메인 규칙은 answer 내부에서 해결해야하기 때문에 현재 방식을 유지해야하는 점
2. 어차피 service layer 에서 트랜잭셔널 범위이기 때문에 예외가 전파되면 롤백되는 점

만약 변경했다면 이렇게 했을 것 같습니다
```java
// Answers 내부
public List<DeleteHistory> deleteAll(NsUser loginUser) throws CannotDeleteException {
    validateAllAuthority(loginUser);  // 1. 모두 권한 있는지 확인
    deleteAllAnswers();               // 2. 그 다음 삭제
    return addInDeleteHistory();
}

private void validateAllAuthority(NsUser loginUser) throws CannotDeleteException {
    for (Answer answer : answers) {
        answer.validateAuthority(loginUser);
    }
}

private void deleteAllAnswers() {
    for (Answer answer : answers) {
        answer.delete(); // 권한은 이미 다 체크됐으니 상태 변경만
    }
}
```
참고부탁드리겠습니다

감사합니다!

javajigi commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -46,34 +40,50 @@ public Answer(Long id, NsUser writer, Question question, String contents) {
+        }
+    }
+    
+    public boolean isNotOwner(NsUser writer) {
```
```suggestion
    private boolean isNotOwner(NsUser writer) {
```
부정문도 메서드로 추상화한 점 👍

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -1,28 +1,22 @@
+import nextstep.users.domain.NsUser;
 
-public class Question {
+public class Question extends Base{
```
```suggestion
public class Question extends Base {
```

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -1,28 +1,22 @@
-
+    
     private String title;
-
+    
     private String contents;
```
지난 수업 중에 다룬 것처럼 필드의 수를 줄이기 위해 이 두 값을 합쳐 QuestionBody와 같은 객체로 분리하는 것은 의미있을가?

javajigi reviewed (approved):
피드백 반영 넘 잘 했네요. 👍 
pr 본문에 남긴 것처럼 exception이 발생할 경우 자동 롤백에 되기 때문에 지금과 같이 구현해도 문제 없습니다.
그 부분을 깨달아 지금과 같이 구현한 점 👍 
1단계 미션 마무리하고 본격적으로 2단계 미션 진행해 보시죠.
